### PR TITLE
add "from" to error and legacy error in the report

### PR DESCRIPTION
**Pull request by OsmondJiang**
https://dev.azure.com/ceapex/Engineering/_workitems/edit/106830

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/4896)

### CODE REVIEW

File: src/docfx/docset/FilePath.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        /// </summary>
+        public string Path { get; }
+
+        /// <summary>
+        /// Gets the value to indicate where is this file from
+        /// </summary>
+        public FileFrom From { get; }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static implicit operator string(FilePath filePath) => filePath?.Path;
```
Comment (yufeih) [inline src/docfx/docset/FilePath.cs]: consider remove this implicit conversion

File: src/docfx/lib/log/ErrorLog.cs [hunk 2]
```diff
@@ -69,7 +69,7 @@ public bool Write(IEnumerable<Error> errors)
             return hasErrors;
         }
 
-        public bool Write(string file, IEnumerable<Error> errors)
+        public bool Write(Document file, IEnumerable<Error> errors)
```
Comment (yufeih) [inline src/docfx/lib/log/ErrorLog.cs]:
```suggestion
        public bool Write(FilePath file, IEnumerable<Error> errors)
```

File: src/docfx/docset/FilePath.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            return Equals(obj as FilePath);
+        }
+
+        public override int GetHashCode()
+        {
+            return HashCode.Combine(PathUtility.PathComparer.GetHashCode(Path), From);
+        }
+
+        public bool Equals(FilePath other)
```
Comment (yufeih) [inline src/docfx/docset/FilePath.cs]: implement == and !=

File: src/docfx/lib/log/Error.cs [hunk 4]
```diff
@@ -21,6 +21,8 @@ internal class Error
 
         public string File { get; }
 
+        public FileFrom? From { get; }
```
Comment (yufeih) [inline src/docfx/lib/log/Error.cs]:
```suggestion
        public FileOrigin Origin { get; }
```

File: src/docfx/docset/FilePath.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        public override int GetHashCode()
+        {
+            return HashCode.Combine(PathUtility.PathComparer.GetHashCode(Path), Origin);
+        }
+
+        public bool Equals(FilePath other)
+        {
+            return other?.Path == Path && other?.Origin == Origin;
```
Comment (yufeih) [inline src/docfx/docset/FilePath.cs]: with this implementation, `new FilePath() == null`

File: src/docfx/docset/FilePath.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        /// The file relative path
+        /// </summary>
+        public string Path { get; }
+
+        /// <summary>
+        /// Gets the value to indicate where is this file from
+        /// </summary>
+        public FileOrigin Origin { get; }
+
+        public FilePath(string path, Docset docset)
```
Comment (yufeih) [inline src/docfx/docset/FilePath.cs]: This class will become a very primitive type to be consumed by a lot of places, so consider:
- move this file to `lib`
- remove the dependency on `Docset`. you can add an `Origin` property to docset if this method is to be shared.

File: src/docfx/lib/PathUtility.cs [hunk 7]
```diff
@@ -262,6 +262,9 @@ public static string UrlToShortName(string url)
             return result.ToString();
         }
 
+        public static bool EndsWith(this FilePath filePath, string value, StringComparison stringComparison)
```
Comment (yufeih) [inline src/docfx/lib/PathUtility.cs]: nit: consider move this method to `FilePath.cs`

File: src/docfx/docset/FilePath.cs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        public bool Equals(FilePath other)
+        {
+            return other?.Path == Path && other?.Origin == Origin;
+        }
+
+        public int CompareTo(FilePath other)
+        {
+            var result = string.Compare(Path, other?.Path, PathUtility.PathComparison);
```
Comment (yufeih) [inline src/docfx/docset/FilePath.cs]: consider `new FilePath vs null` case here.

**Comment by OsmondJiang** [comment]
The diff are expected.

**Pull request merged**